### PR TITLE
refactor: allow ts-node to be scoped

### PR DESCRIPTION
**Pull request by elevatebart**
Advances UNIFY-1216

### TLDR

- This PR allows, in development, to register ts-node on a reduced scope.

### Details

- We want the child process for running `cypress.config.js` to be written in TypeScript.
- Some of the TypeScript code run in this process might not be in Cypress since it's the users.
- We need for ts-node to register once for Cypress's code and once for User's code so that if the 2 configs are different they do not conflict
- This PR does not change the way everything is run in production. We only use ts-node there for the User's code.
- This PR allows, in development, to register ts-node on a reduced scope.
- Then users can use their own `tsconfig.json` for their own files

### CODE REVIEW

File: packages/ts/registerDir.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+const debug = require('debug')('cypress:ts')
+const path = require('path')
+
+// in development we should have TypeScript hook installed
+// in production or staging we are likely to be running
+// built Electron app without ts-node hook. Assume the
+// build has been done correctly
+module.exports = function (scopeDir) {
+  try {
+  // Prevent double-compiling if we're testing the app and already have ts-node hook installed
```
Comment (ryanthemanuel) [inline packages/ts/registerDir.js:11]: Comment spacing off here

Comment (elevatebart) [reply packages/ts/registerDir.js:11]: What do you mean?

File: packages/ts/registerDir.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        scopeDir,
+      })
+    } else {
+      debug('skipping ts-node registration while testing the app')
+    }
+
+  // do we need to prevent any other TypeScript hooks?
+  } catch (e) {
+  // continue running without TypeScript require hook
+    debug('Running without ts-node hook in environment "%s"', process.env.CYPRESS_INTERNAL_ENV)
```
Comment (ryanthemanuel) [inline packages/ts/registerDir.js:38]: Is it worth debug logging the error here?

Comment (elevatebart) [reply packages/ts/registerDir.js:38]: If you are debugging ts, it might be helpful to know if tsconfig is not compatible with ts-node, for instance.

File: packages/ts/register.js [hunk 3]
```diff
@@ -1,34 +1 @@
...
-      preferTsExts: true, // Helps when the files are compiled locally, resolves the TS file
-    })
-  }
-
-  // do we need to prevent any other TypeScript hooks?
-} catch (e) {
-  // continue running without TypeScript require hook
-  log('Running without ts-node hook in environment "%s"', process.env.CYPRESS_INTERNAL_ENV)
-}
+require('./registerDir')()
```
Comment (flotwig) [inline packages/ts/register.js:1]: this could probably use a comment for why it's needed, it wasn't immediately clear how this differed from registerDir to me. is this comment correct?

```suggestion
// This file can be used as a shortcut to invoke `registerDir` without a scope.
// This is useful in Mocha tests with the `-r` option (`-r @packages/ts/register`)
require('./registerDir')()
```

File: packages/config/index.js [hunk 4]
```diff
@@ -1,5 +1,5 @@
 if (process.env.CYPRESS_INTERNAL_ENV !== 'production') {
-  require('@packages/ts/register')
+  require('@packages/ts/registerDir')(__dirname)
```
Comment (flotwig) [inline packages/config/index.js:2]: doesn't this same change need to be made everywhere we use `require('@packages/ts/register')` for full benefits?

**Review by flotwig — CHANGES_REQUESTED** [review]
nice find in ts-node!

File: packages/config/index.js (hunk 4 above)
Comment (elevatebart) [reply packages/config/index.js:2]: Created ticket UNIFY-1548 to apply this to all packages.

**Review by flotwig — APPROVED** [review]

**Review by lmiller1990 — APPROVED** [review]
